### PR TITLE
Refactor Events Package

### PR DESCRIPTION
**Pull request by htdvisser**
#### Summary

This PR refactors the events package to be more efficient, especially when using the Redis backend.

This is related to, but does not yet address https://github.com/TheThingsNetwork/lorawan-stack/issues/3221.

#### Changes

- Change Subscriber interface to take entity identifiers
- Replace the "default" pubsub with a "noop" pubsub
- Implement a new "basic" pubsub
- Change Redis pubsub backend to use multiple channels, and only subscribe to Redis channels when there are actual subscriptions
- Change Redis pubsub backend to use protobuf instead of json

#### Testing

- [x] Kept the existing unit tests for the existing backends as much the same as possible
- [x] Monkey testing with CLI and Console
- [x] Will try to write a common test suite that we can use to test all backend implementations.

##### Regressions

- The most common use of events is the event stream in the Console, and I've tested that.

#### Notes for Reviewers

I'm afraid this is a bit too late for v3.11, even though it would have been nice to get it in there. What we could do, is keep the existing `redis` backend, and instead add the current Redis implementation as a separate backend. Then we can put it behind a config flag, and we could still add it in v3.11.1.

#### Checklist

- [ ] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
	- [ ] Needs minor bump
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: pkg/events/redis/redis.go [hunk 1]
```diff
@@ -17,67 +17,98 @@ package redis
...
-	sub := ps.client.Subscribe(ctx, ps.eventChannel)
-	logger.Info("Subscribed")
+
+	ps.mu.Lock()
+	ps.sub = ps.client.Subscribe(ctx)
+	patterns := make([]string, len(ps.subscriptions))
+	for pattern := range ps.subscriptions {
+		patterns = append(patterns, pattern)
+	}
+	logger.WithField("patterns", patterns).Debug("PSUBSCRIBE")
```
Comment (adriansmares) [inline pkg/events/redis/redis.go]: Maybe the messages should be a bit more descriptive ?

File: pkg/events/redis/redis.go [hunk 2]
```diff
@@ -86,28 +117,108 @@ func (ps *PubSub) subscribeTask(ctx context.Context) error {
...
+
+		ps.PubSub.RemoveSubscription(sub)
+
+		ps.mu.Lock()
+		defer ps.mu.Unlock()
+
+		for _, pattern := range sub.patterns {
+			ps.subscriptions[pattern]--
+			if ps.subscriptions[pattern] == 0 {
+				log.FromContext(ctx).WithField("pattern", pattern).Debug("PUNSUBSCRIBE")
```
Comment (adriansmares) [inline pkg/events/redis/redis.go]: Same here.

**Review by adriansmares — APPROVED** [review]

**Review by johanstokking — APPROVED** [review]
Nothing to add

**Comment by htdvisser** [comment]
Re-requesting review from @adriansmares for the added test suite

**Review by adriansmares — APPROVED** [review]

**Pull request merged**